refactor: Use smart pointers for Simulator, Graphics and KDTree instances

ameysutavani opened the pull request:
- Use smart pointers for Simulator, Graphics and KDTree instances instead of raw pointers
- Make the SEED `unsigned int` to match `srand(unsigned init)` and avoid signed conversions
- Ignore build-* directories as well (usually created by vscode. eg build-vscode)
- Use default dtors instead of user-defined dtors for Robot, Graphics and Simulator classes to clean up resources using RAII (Resource Acquisition Is Initialization)

ameysutavani committed:
Use unsigned init for SEED; Use unique_ptr for Simulator instance

ameysutavani committed:
Ignore build-* folders

ameysutavani committed:
Use unique_ptr for Graphics and KDTree

ameysutavani commented:
@aalpatya Couple you take a look at this one as well when you get a chance? This should be a much more straightforward set of changes to review.

ameysutavani commented:
Hi @aalpatya,
Could you please take a look at this PR when possible?

aalpatya reviewed (commented):
Great work, however when gbpplanner is running, if you press [ESC], I run into a seg fault. Do you get this?
I think it's to do with the destructor of ~Simulator().

Running with debug in vs code and stepping through, the seg fault happens during the destructor of the Graphics class. 

Hope this screenshot displays on git:
![image](https://github.com/user-attachments/assets/459729f3-272e-4152-92c2-42c10269f00e)

aalpatya reviewed (commented):
Deleting (reset() ) the graphics pointer *before* the call to Close the window seems to work.
But I haven't used unique_ptr before, so not sure if this is the right thing to do.

I think what was happening was that raylib's CloseWindow was executing before the Graphics class tried to delete the loaded texture (sort of like pulling the rug from underneath you).

* Could you investigate a little bit more if you have time and then make the change? Thanks!

![image](https://github.com/user-attachments/assets/6a69deff-6fcb-461d-8ee0-e115265749a5)

ameysutavani commented:
Thanks for testing this @aalpatya. I will replicate this behavior on my end and post my findings.

ameysutavani committed:
Use default dtors for Robot, Graphics, Simulator classes

ameysutavani commented:
@aalpatya You were right. The Graphics class's user-defined dtor calls UnloadTexture() of raylib. This needs the Window to be present to clear the texture from. I have added explicit graphics.reset() call before we call CloseWindow(). I defaulted the dtors for the Graphics and the Simulator class so that they release all the resources of the classes in the reverse order of their definition. We also now don't need a call to UnloadTexture() and CloseWindow() as the raylib instance inside the Simulator class will be automatically released when the default ~Simulator is call as we go out of the main().
I also changed the order in which the robot kdtree is defined in the ctor, so that it gets cleaned first.

TL;DR: RAII works correctly through all classes now

I have tested this myself with the debugger; can you please also test once yourself?

aalpatya commented on src/Simulator.cpp:
```diff
@@ -19,30 +19,17 @@ Simulator::Simulator(){
-        CloseWindow();
-    }
+    // Initialise kdtree for storing robot positions (needed for nearest neighbour check)
+    treeOfRobots_ = std::make_unique<KDTree>(2, robot_positions_, 50); 
```
Why is this inside the destructor?

ameysutavani replied:
Oh, the diff is making it look like this. This line is inside the constructor itself: https://github.com/aalpatya/gbpplanner/blob/b167a78ee907af72793bbaafb21914b186cacba9/src/Simulator.cpp#L32

aalpatya commented on src/Graphics.cpp:
```diff
@@ -61,10 +61,6 @@ Graphics::Graphics(Image obstacleImg) : obstacleImg_(ImageCopy(obstacleImg)){
 }
 
-Graphics::~Graphics(){
-    UnloadTexture(texture_img_);
```
I believe I had added the UnloadTexture in the case that a background image was given to the Graphics class
(eg. for obstacle avoidance, the obstacles image was loaded as the background). Then, on exit, according to Raylib you have to Unload the texture.
Correct me if this is not needed anymore

ameysutavani replied:
I think in our Simulator's case, when the ESC is hit, the entire program dies and goes out of scope. The respective default destructors of the smart pointers should clean-up the allocations in the reverse order of their allocations. But I can run asan just for a sanity check.

We have the first case from here: https://github.com/raysan5/raylib/discussions/4200#discussioncomment-10965628

ameysutavani replied:
Thinking twice about it, I am going to call UnloadTexture() to err on the side of caution.

https://github.com/RobLoach/raylib-cpp might be be what we need for the future iterations

ameysutavani committed:
Keep the user-defined destructors for Simulator and Graphics classes for now

ameysutavani commented:
@aalpatya Reverted back to have user-defined destructors for Simulator and the Graphics classes. I think I will issue a follow-up PR to switch to raylib-cpp which would reduce us having to manage raylib's memory (should be easy). Would love to hear your thoughts on it before.
